File: possible_moves.py
from typing import List, Sequence, Tuple, Callable, Generator
from classes import Piece, PieceWithPositions, GameBoard
from functools import partial


def get_initial_game_board() -> GameBoard:

    def get_Piece_with_right_Team(row: int, column: int):
        team = "R" if (row + column) % 2 == 0 and row < 3 else \
            "G" if (row + column) % 2 == 0 and row > 4 else " "
        return Piece(position=(row, column), team=team, king=False)

    return GameBoard(game_board=tuple(
        tuple(get_Piece_with_right_Team(row, column) for column in range(8))
        for row in range(8)), currPlayer="G")


def get_possible_move_vectors(game_board: GameBoard, piece: Piece, is_piece_there: bool) -> Tuple[Tuple[int, int], ...]:
    if is_piece_there:
        vectors_which_piece_can_move = (
            (1, -1), (1, 1)) if piece.team == "R" else ((-1, -1), (-1, 1))
        if piece.king == True:
            vectors_which_piece_can_move = ((1, -1), (1, 1), (-1, -1), (-1, 1))
        return vectors_which_piece_can_move
    return tuple()


def filter_without_smash(game_board: GameBoard, position: Tuple[int, int]):
    in_bound = check_if_position_out_of_bound(position)
    return in_bound and not is_piece_there_on_row_column(
        game_board=game_board, piece=game_board.game_board[position[0]][position[1]]
    )


def filter_with_smash(game_board: GameBoard, piece: Piece, vector: Tuple[int, int]):
    def opposite_pieces_on_position(own_piece: Piece, newPositionPiece: Piece):
        return own_piece.team != " " and newPositionPiece.team != " " and own_piece.team != newPositionPiece.team
    row, column = piece.position
    row_vector, column_vector = vector

    in_bound_intermediate = check_if_position_out_of_bound(
        (row + row_vector, column + column_vector))
    in_bound_destination = check_if_position_out_of_bound(
        (row + 2*row_vector, column + 2*column_vector))
    if in_bound_intermediate and in_bound_destination:
        opposite = opposite_pieces_on_position(
            piece, game_board.game_board[row + row_vector][column + column_vector])
        without_piece_on_position = not is_piece_there_on_row_column(
            game_board=game_board, piece=game_board.game_board[row+2*row_vector][column+2*column_vector])
        return opposite and without_piece_on_position
    else:
        return False


def get_possible_moves(game_board: GameBoard, piece: Piece,
                       vectors_which_piece_can_move: List[Tuple[int, int]],
                       smash: bool, filter_function: Callable) -> Tuple[Tuple[int, int], ...]:
    row, column = piece.position
    if smash:
        possible_positions = tuple((row + 2*row_vector, column + 2*column_vector)
                                   for row_vector, column_vector in vectors_which_piece_can_move
                                   if filter_function(game_board=game_board, piece=piece, vector=(row_vector, column_vector)))
    else:
        possible_positions = tuple((row + row_vector, column + column_vector)
                                   for row_vector, column_vector in vectors_which_piece_can_move if filter_function(game_board=game_board, position=(row + row_vector, column + column_vector)))
    return possible_positions


def get_possible_smash_moves(game_board: GameBoard, piece: Piece,
                             vectors_which_piece_can_move: List[Tuple[int, int]]
                             ) -> Tuple[Tuple[int, int], ...]:
    return get_possible_moves(game_board=game_board, piece=piece, vectors_which_piece_can_move=vectors_which_piece_can_move, smash=True, filter_function=filter_with_smash)


def get_possible_moves_without_smash(game_board: GameBoard, piece: Piece,
                                     vectors_which_piece_can_move: List[Tuple[int, int]]):
    return get_possible_moves(game_board=game_board, piece=piece, vectors_which_piece_can_move=vectors_which_piece_can_move, smash=False, filter_function=filter_without_smash)


def check_if_position_out_of_bound(
    position: Tuple[int, int]): return 0 <= position[0] < 8 and 0 <= position[1] < 8


def is_piece_there_on_row_column(game_board: GameBoard, piece: Piece): return check_if_position_out_of_bound(
    piece.position) and (game_board.game_board[piece.position[0]][piece.position[1]].team != " ")


def get_all_possible_moves_for_piece(game_board: GameBoard, piece: Piece, possible_move_function: Callable) \
        -> Piece:

    vectors_which_piece_can_move = get_possible_move_vectors(
        game_board=game_board, piece=piece, is_piece_there=is_piece_there_on_row_column(game_board=game_board, piece=piece))
    if vectors_which_piece_can_move:
        possible_postions = possible_move_function(
            game_board=game_board, piece=piece,
            vectors_which_piece_can_move=vectors_which_piece_can_move)
        if possible_postions:
            return PieceWithPositions(**piece.model_dump(), possible_positions=possible_postions)
        else:
            return piece
    else:
        return piece


def get_all_possible_moves_for_piece_gui(game_board: GameBoard, piece) -> List[Tuple[int, int]]:
    all_moves = get_all_possible_moves_for_team(
        game_board=game_board, team=piece.team)
    potential_move = next(filter(lambda piece_iter: isinstance(piece_iter, PieceWithPositions)
                                 and piece_iter.position == piece.position, all_moves), None)
    if all_moves and potential_move:
        return potential_move.possible_positions
    else:
        return []


def get_all_possible_moves_for_team(game_board: GameBoard, team: str) -> List[Piece]:

    team_pieces = list(filter(lambda piece: piece.team ==
                              team, (piece for row in game_board.game_board for piece in row)))
    pieces_with_possible_smash = map(lambda piece: get_all_possible_moves_for_piece(
        game_board=game_board, piece=piece,
        possible_move_function=get_possible_smash_moves), team_pieces.copy())

    pieces_with_possible_smash_moves = list(filter(lambda piece: isinstance(
        piece, PieceWithPositions), pieces_with_possible_smash))

    if not pieces_with_possible_smash_moves:
        pieces_with_possible_moves_without_smash = map(lambda piece: get_all_possible_moves_for_piece(
            game_board=game_board, piece=piece,
            possible_move_function=get_possible_moves_without_smash), team_pieces)
        return list(filter(lambda piece: isinstance(piece, PieceWithPositions), pieces_with_possible_moves_without_smash))
    else:
        return pieces_with_possible_smash_moves

File: Go/possible_moves/possible_moves.go
package possible_moves

import (
	. "checkers/packages/gameboard"

	"github.com/samber/lo"
)

type PossibleMoveFunction func(gameBoard GameBoard, piece Piece, vectorsWhichPieceCanMove []Tuple) []Tuple

type FilterFunction func(gameBoard GameBoard, piece Piece, vector Tuple) bool

func checkIfPositionOutOfBounds(position Tuple) bool {
	return 0 <= position.Row && position.Row < 8 && 0 <= position.Column && position.Column < 8
}

func isPieceThereOnRowColumn(piece Piece) bool {
	return checkIfPositionOutOfBounds(piece.Position) && piece.Team != " "
}

func GetPossibleMoveVectors(piece Piece, isPieceThere bool) []Tuple {
	vectorsWhichPieceCanMove := []Tuple{}
	if isPieceThere && piece.Team != " " {
		if piece.Team == "R" {
			vectorsWhichPieceCanMove = []Tuple{{Row: 1, Column: -1}, {Row: 1, Column: 1}}
		} else if piece.Team == "G" {
			vectorsWhichPieceCanMove = []Tuple{{Row: -1, Column: -1}, {Row: -1, Column: 1}}
		}
		if piece.King {
			vectorsWhichPieceCanMove = []Tuple{{Row: -1, Column: -1}, {Row: -1, Column: 1}, {Row: 1, Column: -1}, {Row: 1, Column: 1}}
		}
	}
	return vectorsWhichPieceCanMove
}

func FilterWithoutSmash(gameBoard GameBoard, piece Piece, vector Tuple) bool {
	row, column := piece.Position.Row, piece.Position.Column
	rowVector, columnVector := vector.Row-row, vector.Column-column
	position := Tuple{Row: row + rowVector, Column: column + columnVector}
	inBound := checkIfPositionOutOfBounds(position)
	return inBound && !isPieceThereOnRowColumn(gameBoard.GameBoard[position.Row][position.Column])
}

func FilterWithSmash(gameBoard GameBoard, piece Piece, vector Tuple) bool {
	oppositePiecesOnPosition := func(ownPiece, newPositionPiece Piece) bool {
		return ownPiece.Team != " " && newPositionPiece.Team != " " && ownPiece.Team != newPositionPiece.Team
	}

	row, column := piece.Position.Row, piece.Position.Column
	rowVector, columnVector := vector.Row-row, vector.Column-column
	intermediate_row, intermediate_column := row+int(float64(rowVector)*0.5), column+int(float64(columnVector)*0.5)

	inBoundIntermediate := checkIfPositionOutOfBounds(Tuple{Row: intermediate_row, Column: intermediate_column})
	inBoundDestination := checkIfPositionOutOfBounds(Tuple{Row: row + rowVector, Column: column + columnVector})

	if inBoundIntermediate && inBoundDestination {
		opposite := oppositePiecesOnPosition(piece, gameBoard.GameBoard[intermediate_row][intermediate_column])
		withoutPieceOnPosition := !isPieceThereOnRowColumn(gameBoard.GameBoard[row+rowVector][column+columnVector])
		return opposite && withoutPieceOnPosition
	} else {
		return false
	}
}

func GetPossibleMoves(smash bool, filterFunction func(GameBoard, Piece, Tuple) bool) func(gameBoard GameBoard, piece Piece, vectorsWhichPieceCanMove []Tuple) []Tuple {
	return func(gameBoard GameBoard, piece Piece, vectorsWhichPieceCanMove []Tuple) []Tuple {
		row, column := piece.Position.Row, piece.Position.Column
		var possiblePositions []Tuple
		map_function := func(smash bool, row int, column int) []Tuple {
			factor := 1
			if smash {
				factor = 2
			}
			return lo.Map(vectorsWhichPieceCanMove, func(vector Tuple, _ int) Tuple {
				rowVector, columnVector := vector.Row, vector.Column
				return Tuple{Row: row + factor*rowVector, Column: column + factor*columnVector}
			})
		}
		possiblePositions = lo.Filter(map_function(smash, row, column), func(vector Tuple, _ int) bool {
			return filterFunction(gameBoard, piece, vector)
		})

		return possiblePositions

	}
}

func GetAllPossibleMovesForPiece(gameBoard GameBoard, piece Piece, possibleMoveFunction PossibleMoveFunction, filterFunction FilterFunction) Piece {
	vectorsWhichPieceCanMove := GetPossibleMoveVectors(piece, isPieceThereOnRowColumn(piece))
	if len(vectorsWhichPieceCanMove) > 0 {
		possiblePositions := possibleMoveFunction(gameBoard, piece, vectorsWhichPieceCanMove)
		piece.PossiblePositions = possiblePositions
	}
	return piece
}

func GetAllPossibleMovesForTeam(gameBoard GameBoard, team string) []Piece {
	PossibleMoveFunctionWithSmash := GetPossibleMoves(true, FilterWithSmash)
	PossibleMoveFunctionWithoutSmash := GetPossibleMoves(false, FilterWithoutSmash)
	map_function := func(possibleMoveFunction PossibleMoveFunction, piece Piece) Piece {
		smashes := possibleMoveFunction(gameBoard, piece, GetPossibleMoveVectors(piece, isPieceThereOnRowColumn(piece)))
		piece.PossiblePositions = smashes
		return piece
	}
	filter_function := func(piece Piece, _ int) bool {
		return len(piece.PossiblePositions) > 0
	}

	flatten_board := lo.Flatten(gameBoard.GameBoard)
	team_pieces := lo.Filter(flatten_board, func(piece Piece, _ int) bool {
		return piece.Team == team
	})
	team_pieces_with_smash_positions := lo.Map(team_pieces, func(piece Piece, _ int) Piece {
		return map_function(PossibleMoveFunctionWithSmash, piece)
	})
	true_smashes := lo.Filter(team_pieces_with_smash_positions, filter_function)
	if len(true_smashes) == 0 {
		team_pieces_with_possible_positions := lo.Map(team_pieces, func(piece Piece, _ int) Piece {
			return map_function(PossibleMoveFunctionWithoutSmash, piece)
		})
		team_pieces_with_possible_positions = lo.Filter(team_pieces_with_possible_positions, filter_function)
		return team_pieces_with_possible_positions
	} else {
		return true_smashes
	}
}

func Make_move(gameBoard GameBoard, action Action) GameBoard {
	piece := gameBoard.GameBoard[action.Start.Row][action.Start.Column]
	king := isPieceCheckerAfterMove(piece, action)
	smash := isOppositePieceSmashed(action)
	new_game_board := MakeNewGameBoardAfterMove(gameBoard, action, smash, king)
	return new_game_board
}
func GetActionsFromPossibleMoves(gameBoard GameBoard, possibleMoves []Piece) []Action {

	convertPieceToAction := func(piece Piece, possiblePosition Tuple) Action {
		return Action{Start: piece.Position, End: possiblePosition}
	}
	getActionFromOnePiece := func(piece Piece, _ int) []Action {
		return lo.Map(piece.PossiblePositions, func(possiblePosition Tuple, _ int) Action {
			return convertPieceToAction(piece, possiblePosition)
		})
	}
	actions := lo.Map(possibleMoves, getActionFromOnePiece)
	actions_flattened := lo.Flatten(actions)
	return actions_flattened
}

func isOppositePieceSmashed(action Action) bool {
	abs := func(x int) int {
		if x < 0 {
			return -x
		}
		return x
	}
	oldRow, oldColumn := action.Start.Row, action.Start.Column
	row, column := action.End.Row, action.End.Column
	if abs(row-oldRow) == 2 || abs(column-oldColumn) == 2 {
		return true
	}
	return false
}

func isPieceCheckerAfterMove(piece Piece, action Action) bool {
	row := action.End.Row
	if (piece.Team == "R" && row == 7) || (piece.Team == "G" && row == 0) || piece.King {
		return true
	}
	return false
}
